core/model-api/src/main/java/org/eclipse/rdf4j/model/base/AbstractIRI.java

```diff
@@ -34,17 +33,62 @@ public String stringValue() {
 	@Override
 	public boolean equals(Object o) {
 		return this == o || o instanceof IRI
-				&& Objects.equals(toString(), ((Value) o).toString()); // !!! use stringValue()
+				&& toString().equals(o.toString()); // !!! use stringValue()
 	}
 
 	@Override
 	public int hashCode() {
-		return Objects.hashCode(toString()); // !!! use stringValue()
+		return toString().hashCode(); // !!! use stringValue()
 	}
 
 	@Override
 	public String toString() {
 		return stringValue();
 	}
 
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	static class GenericIRI extends AbstractIRI {
+
+		private static final long serialVersionUID = 2209156550690548467L;
+
+		private String iri;
+
+		private int split;
+
+		GenericIRI(String iri) {
+			this.iri = iri;
+			this.split = 0;
+		}
+
+		GenericIRI(String namespace, String localName) {
+			this.iri = namespace + localName; // !!! concatenation performance impact?
+			this.split = namespace.length();
+		}
+
+		@Override
+		public String stringValue() {
+			return iri;
+		}
+
+		@Override
+		public String getNamespace() {
+			return iri.substring(0, split());
+		}
+
+		@Override
+		public String getLocalName() {
+			return iri.substring(split());
+		}
+
+		private int split() {
+			return (split > 0) ? split
+					: (split = iri.indexOf('#') + 1) > 0 ? split
+							: (split = iri.lastIndexOf('/') + 1) > 0 ? split
+									: (split = iri.lastIndexOf(':') + 1) > 0 ? split
+											: 0; // unexpected: colon presence already tested in factory methods
```

Comment:
Could we rewrite this using if...else instead of a collection of ternary operators? It will be easier to read I think.

Reply:
Done